engineering-plans/0002-subgraph-schema-merging.md

```diff
@@ -0,0 +1,118 @@
+# PLAN-0001: Subgraph Schema Merging for Subgraph Composition
+
+<dl>
+  <dt>Author</dt>
+  <dd>Jorge Olivero</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0001-subgraph-composition.md">RFC-0001 Subgraph Composition</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/5">Engineering Plan PR</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-09</dd>
+
+  <dt>Date of approval</dt>
+  <dd>TBD</dd>
+
+  <dt>Approved by</dt>
+  <dd>TBD</dd>
+</dl>
+
+## Summary
+
+Subgraph composition allows a subgraph to import types from a foreign subgraph. Imports are designed to be loosely coupled throughout the deployment and indexing phases, and tightly coupled during query execution for the subgraph.
+
+To generate an API schema for the subgraph, the local schema needs to be merged with each of the foreign schemas that it imports. The merging process needs to take the following into account:
+
+1. An imported schema not being available on the graph-node
+2. An imported schema without a definition for the type which the local subgraph imports
+3. A merged and cached api schema for a subgraph imported by name is updated
+4. Ability to tell which subgraph/schema each type belongs to
+
+## Implementation
+
+There are two important parts to schema merging:
+
+1. Merging the schema
+
+2. Invalidating a cache entry when a subgraph's dependency, referenced by name, is updated and remerging that schema
```

Comment:
Nothing above speaks about caching yet, so perhaps we should talk about that first? What is cached and why and how?

Reply:
So I'd say something like
```
## Implementation

The schema merging implementation consists of two parts:

1. A cache of subgraph schemas
  - Each schema is associated with a subgraph name and/or deployment ID
  - Each schema is associated with all its dependencies (subgraph names and/or deployment IDs)
  - Each schema is stored in its merged form
2. The actual schema merging logic

I'm making assumptions about the cache here, mainly because they are not specified well enough in this plan yet.